src/org/ods/orchestration/FinalizeStage.groovy

```diff
@@ -115,4 +124,107 @@ class FinalizeStage extends Stage {
+        def gatherCommitTasks = flattenedRepos.collectEntries { repo ->
+            [
+                (repo.id): {
+                    steps.dir("${steps.env.WORKSPACE}/${MROPipelineUtil.REPOS_BASE_DIR}/${repo.id}") {
+                        repo.data.git.createdExecutionCommit = git.commitSha
+                    }
+                }
+            ]
+        }
+        gatherCommitTasks.failFast = true
+        script.parallel(gatherCommitTasks)
+    }
+
+    private void integrateIntoMainBranch(IPipelineSteps steps, GitService git) {
+        def flattenedRepos = repos.flatten()
+        def repoIntegrateTasks = flattenedRepos.collectEntries { repo ->
+            [
+                (repo.id): {
+                    steps.dir("${steps.env.WORKSPACE}/${MROPipelineUtil.REPOS_BASE_DIR}/${repo.id}") {
+                        def filesToCheckout = []
+                        if (steps.fileExists('openshift')) {
+                            filesToCheckout = ['openshift/ods-deployments.json']
+                        } else {
+                            filesToCheckout = [
+                                'openshift-exported/ods-deployments.json',
+                                'openshift-exported/template.yml'
+                            ]
+                        }
+                        git.mergeIntoMainBranch(project.gitReleaseBranch, repo.branch, filesToCheckout)
+                    }
+                }
+            ]
+        }
+        repoIntegrateTasks.failFast = true
+        script.parallel(repoIntegrateTasks)
+    }
+
+    private void recordAndPushEnvState(IPipelineSteps steps, ILogger logger, GitService git) {
+        // record release manager repo state
+        logger.debug "Finalize: Recording HEAD commits from repos ..."
+        logger.debug "On release manager commit ${git.commitSha}"
+        def gitHeads = repos.flatten().collectEntries { repo ->
+            logger.debug "HEAD of repo '${repo.id}': ${repo.data.git.createdExecutionCommit}"
+            [(repo.id): (repo.data.git.createdExecutionCommit ?: '')]
+        }
+        def envState = [
+            version: project.buildParams.version,
+            changeId: project.buildParams.changeId,
+            repositories: gitHeads
+        ]
+        steps.writeFile(
+            file: project.envStateFileName,
+            text: JsonOutput.prettyPrint(JsonOutput.toJson(envState))
+        )
+        git.commit(
+            [project.envStateFileName],
+            "ODS: Record commits deployed into ${project.buildParams.targetEnvironmentToken}"
+        )
+
+        if (project.isWorkInProgress) {
+            git.pushRef('master')
+        } else {
+            // We don't need to merge, we simply commit the env file. That
+            // avoids unnecessary merge conflicts.
+            git.switchToOriginTrackingBranch('master')
+            git.checkoutAndCommitFiles(
+                project.gitReleaseBranch,
+                [project.envStateFileName],
```

Comment:
In case we need more files to be merged back in the MRO repo, this should be the place to add it, right?

Reply:
In theory, yes. But note that I'm actually *not* merging the release branch of the release manager repository back into the `master` branch. I'm directly committing one file (the env state) on `master`. This is done to avoid merge conflicts. Please check if you actually want to have a merge of the release manager ... I didn't do it out of fear it might create unnecessary merge conflicts (e.g. in `dev` state) ... and at the same time, I did not see a use case where one really needs the merge back into the main branch.

Reply:
Yes, in fact this sort of works with the plan I had to store the jira data for a given version. Saving one file per version to avoid potential merge conflicts as much as possible. So actually it looks like adding the file there would be enough to bring the information to master branch. Great!